### PR TITLE
max transaction cost in mempool

### PR DESCRIPTION
**Pull request by pragmaxim**
Addresses https://github.com/ergoplatform/ergo/issues/1504 ... peer which is sending costly transaction should be considered spammy one

### CODE REVIEW

File: src/main/scala/org/ergoplatform/mining/CandidateGenerator.scala [hunk 1]
```diff
@@ -710,7 +711,8 @@ object CandidateGenerator extends ScorexLogging {
             stateWithTxs.validateWithCost(
               tx,
               Some(upcomingContext),
-              maxTransactionComplexity
+              maxTransactionComplexity,
+              maxTransactionCost
             ) match {
```
Comment (pragmaxim) [inline src/main/scala/org/ergoplatform/mining/CandidateGenerator.scala:716]: I think that letting a transaction with cost over limit into a candidate is also wrong

Comment (kushti) [reply src/main/scala/org/ergoplatform/mining/CandidateGenerator.scala:713]: The limit is for mempool, to protect from it spam, and prevent broadcasting too heavy transactions. For blocks, still there's another limit (cost per block) only, and miner can include transactions which exceed maxTransactionCost directly.

**Comment by kushti** [comment]
@pragmaxim merged #1511 into this PR, please do new tests for too heavy transactions instead of ones removed (in #1511)

**Comment by pragmaxim** [comment]
@kushti test changed from `complexity` -> `cost`

**Comment by kushti** [comment]
@pragmaxim replaced tx cost limit in miner (CandidateGenerator) with block cost limit. Miner should have ability to include transactions not propagated directly into the blocks (as verification will be passed anyway). Maybe block cost limit is not very precise though ?

**Comment by pragmaxim** [comment]
@kushti not sure I understand : 
>>> Maybe block cost limit is not very precise though ?

**Comment by kushti** [comment]
> @kushti not sure I understand :
> 
> > > > Maybe block cost limit is not very precise though ?

I meant maybe better to have some gap, for security and also fee collecting tx

**Comment by pragmaxim** [comment]
@kushti I'm a bit lost in this PR ... do you mean hard code it like this :  `maxBlockCost + someExtraReserveCost` ?

**Comment by pragmaxim** [comment]
@kushti please check last 2 commits, I wanted to finish this PR finally

**Comment by kushti** [comment]
> @kushti please check last 2 commits, bug-fix + extraFeeCost ... passing context is important as otherwise it is read from DB

cost in transaction assembly is computational cost (similarly to gas in Ethereum), unlikely it can be correlated to tx fee

**Comment by pragmaxim** [comment]
> cost in transaction assembly is computational cost (similarly to gas in Ethereum), unlikely it can be correlated to tx fee

yeah I misinterpreted that, it is totally unrelated :facepalm:

File: src/main/resources/mainnet.conf [hunk 2]
```diff
@@ -34,6 +34,9 @@ ergo {
 
     # List with hex-encoded identifiers of transactions banned from getting into memory pool
     blacklistedTransactions = []
+
+    # maximum cost of transaction for it to be propagated
+    maxTransactionCost = 5000000
```
Comment (aslesarenko) [inline src/main/resources/mainnet.conf:39]: Looks too high.

File: src/main/resources/testnet.conf [hunk 3]
```diff
@@ -8,10 +8,15 @@ ergo {
...
     headerChainDiff = 800
+
     # having more internal miner threads helps to reproduce real use cases, like 100 GPU miners
     internalMinersCount = 1
+
     # GPU miners are currently polling at 100ms interval
     internalMinerPollingInterval = 200ms
+
+    # maximum cost of transaction for it to be propagated
+    maxTransactionCost = 800000
```
Comment (aslesarenko) [inline src/main/resources/testnet.conf:19]: Why each conf file has its own value for this parameter?

File: src/main/scala/org/ergoplatform/nodeView/state/UtxoStateReader.scala [hunk 4]
```diff
@@ -38,22 +39,23 @@ trait UtxoStateReader extends ErgoStateReader with TransactionValidation {
...
     val verifier = interpreterOpt.getOrElse(ErgoInterpreter(context.currentParameters))
 
     tx.statelessValidity().flatMap { _ =>
       val boxesToSpend = tx.inputs.flatMap(i => boxById(i.boxId))
-      val txComplexity = boxesToSpend.map(_.ergoTree.complexity).sum
-      if (txComplexity > complexityLimit) {
-        throw new Exception(s"Transaction $tx has too high complexity $txComplexity")
-      }
       tx.statefulValidity(
-        boxesToSpend,
```
Comment (aslesarenko) [inline src/main/scala/org/ergoplatform/nodeView/state/UtxoStateReader.scala:54]: validateStateful internally uses MaxBlockCost as the limit. Now, when we have costLimit, it should be passed in validateStateful and take place of MaxBlockCost limit.

File: src/main/scala/org/ergoplatform/nodeView/state/UtxoStateReader.scala [hunk 5]
```diff
@@ -38,22 +39,23 @@ trait UtxoStateReader extends ErgoStateReader with TransactionValidation {
...
-        tx.dataInputs.flatMap(i => boxById(i.boxId)),
-        context)(verifier)
+          boxesToSpend,
+          tx.dataInputs.flatMap(i => boxById(i.boxId)),
+          context)(verifier).flatMap {
+            case txCost if txCost > costLimit =>
```
Comment (aslesarenko) [inline src/main/scala/org/ergoplatform/nodeView/state/UtxoStateReader.scala:54]: At the time we check this limit, the node may already have spent full MaxBlockCost of computations. If the ratio between MaxBlockCost and MaxTxCost become higher, then this is increasingly may be an attack vector. See my previous comment.

**Review by aslesarenko — CHANGES_REQUESTED** [review]
Looks good overall, but see comments.

File: src/main/resources/mainnet.conf (hunk 2 above)
Comment (kushti) [reply src/main/resources/mainnet.conf:39]: Block limit is 7M, and 5M is non-discriminative to most known existing transactions while also limiting them.

File: src/main/resources/testnet.conf (hunk 3 above)
Comment (kushti) [reply src/main/resources/testnet.conf:19]: Block limit is 1M in the testnet, 7M in the mainnet. Thus max tx cost values are different as well.

File: src/main/scala/org/ergoplatform/nodeView/state/UtxoStateReader.scala (hunk 4 above)
Comment (kushti) [reply src/main/scala/org/ergoplatform/nodeView/state/UtxoStateReader.scala:54]: see my reply below

File: src/main/scala/org/ergoplatform/nodeView/state/UtxoStateReader.scala (hunk 5 above)
Comment (kushti) [reply src/main/scala/org/ergoplatform/nodeView/state/UtxoStateReader.scala]: good catch, made a new version of *validateWithCost* which is passing accumulated cost considering the limit and tests, please check

@aslesarenko made 
```case Failure(mme: MalformedModifierError) if mme.message.contains("CostLimitException")``` as no structured information to extract CostLimitException available

File: src/main/scala/org/ergoplatform/nodeView/state/UtxoStateReader.scala [hunk 6]
```diff
@@ -38,22 +40,30 @@ trait UtxoStateReader extends ErgoStateReader with TransactionValidation {
...
-        throw new Exception(s"Transaction $tx has too high complexity $txComplexity")
-      }
       tx.statefulValidity(
-        boxesToSpend,
-        tx.dataInputs.flatMap(i => boxById(i.boxId)),
-        context)(verifier)
+          boxesToSpend,
+          tx.dataInputs.flatMap(i => boxById(i.boxId)),
+          context,
+          accCost)(verifier).map(_ - accCost) match {
```
Comment (aslesarenko) [inline src/main/scala/org/ergoplatform/nodeView/state/UtxoStateReader.scala:58]: The value passed to statefulValidity shouldn't be greater than MaxTxCost limit. 
accCost, as it is computed above, is not bound by MaxTxCost.
So, changes in validateStateful are also necessary.

**Review by aslesarenko — CHANGES_REQUESTED** [review]

File: src/main/scala/org/ergoplatform/nodeView/state/UtxoStateReader.scala (hunk 6 above)
Comment (kushti) [reply src/main/scala/org/ergoplatform/nodeView/state/UtxoStateReader.scala:58]: Cost is bounded by MaxBlockCost, so we set accCost as there's only txCost room to reach it. Please check updated tests.

**Review by oskin1 — APPROVED** [review]

**Pull request merged**